File: killrvideo-service-sugestedvideo/src/test/java/com/killrvideo/service/sugestedvideo/grpc/SuggestedVideosServiceGrpcMapperTest.java
package com.killrvideo.service.sugestedvideo.grpc;

import com.killrvideo.dse.dto.ResultListPage;
import com.killrvideo.dse.dto.Video;
import com.killrvideo.service.sugestedvideo.request.GetRelatedVideosRequestData;
import killrvideo.suggested_videos.SuggestedVideosService.*;
import killrvideo.video_catalog.events.VideoCatalogEvents.*;
import org.junit.jupiter.api.Test;

import java.time.Instant;
import java.util.HashSet;
import java.util.List;
import java.util.Optional;
import java.util.UUID;

import static com.killrvideo.utils.GrpcMappingUtils.*;
import static java.util.Arrays.asList;
import static java.util.Collections.singletonList;
import static org.junit.jupiter.api.Assertions.*;

class SuggestedVideosServiceGrpcMapperTest {
    private final SuggestedVideosServiceGrpcMapper mapper = new SuggestedVideosServiceGrpcMapper();

    @Test
    public void testMapVideoAddedtoVideoDTO() {
        Video v = video();

        YouTubeVideoAdded videoAdded = YouTubeVideoAdded.newBuilder()
                .setVideoId(uuidToUuid(v.getVideoid()))
                .setAddedDate(instantToTimeStamp(v.getAddedDate()))
                .setUserId(uuidToUuid(v.getUserid()))
                .setName(v.getName())
                .addAllTags(v.getTags())
                .setPreviewImageLocation(v.getPreviewImageLocation())
                .setLocation(v.getLocation())
                .build();

        Video video = mapper.mapVideoAddedtoVideoDTO(videoAdded);
        assertEquals(v.getVideoid(), video.getVideoid());
        assertEquals(v.getUserid(), video.getUserid());
        assertEquals(v.getAddedDate(), video.getAddedDate());
        assertEquals(v.getName(), video.getName());
        assertEquals(v.getTags().size(), video.getTags().size());
        assertEquals(v.getPreviewImageLocation(), video.getPreviewImageLocation());
        assertEquals(v.getLocation(), video.getLocation());
    }

    @Test
    public void testMapVideotoSuggestedVideoPreview() {
        Video v = video();
        SuggestedVideoPreview preview = mapper.mapVideotoSuggestedVideoPreview(v);

        assertEquals(v.getVideoid().toString(), preview.getVideoId().getValue());
        assertEquals(v.getUserid().toString(), preview.getUserId().getValue());
        assertEquals(v.getAddedDate(), timestampToInstant(preview.getAddedDate()));
        assertEquals(v.getName(), preview.getName());
        assertEquals(v.getPreviewImageLocation(), preview.getPreviewImageLocation());
    }

    @Test
    public void testParseGetRelatedVideosRequestData() {
        UUID videoid = UUID.randomUUID();
        GetRelatedVideosRequest request = GetRelatedVideosRequest.newBuilder()
                .setVideoId(uuidToUuid(videoid))
                .setPageSize(2)
                .setPagingState("Paging state")
                .build();

        GetRelatedVideosRequestData pojo = mapper.parseGetRelatedVideosRequestData(request);

        assertEquals(videoid, pojo.getVideoid());
        assertEquals(2, pojo.getPageSize());
        assertEquals("Paging state", pojo.getPagingState().get());
    }

    @Test
    public void testMapToGetRelatedVideosResponse() {
        Video v = video();
        UUID videoid = UUID.randomUUID();
        ResultListPage<Video> resultPage = new ResultListPage<>(singletonList(v), Optional.of("next page"));

        GetRelatedVideosResponse response = mapper.mapToGetRelatedVideosResponse(resultPage, videoid);
        assertEquals(videoid.toString(), response.getVideoId().getValue());
        assertEquals("next page", response.getPagingState());
        assertEquals(1, response.getVideosCount());
    }

    private Video video() {
        UUID videoid = UUID.randomUUID();
        UUID userid = UUID.randomUUID();
        Instant addedDate = Instant.now();
        String name = "Game";
        List<String> tagList = asList("tag1", "tag2");
        String previewLocation = "url";
        String location = "locationUrl";
        Video video = new Video();
        video.setVideoid(videoid);
        video.setUserid(userid);
        video.setName(name);
        video.setTags(new HashSet<>(tagList));
        video.setPreviewImageLocation(previewLocation);
        video.setLocation(location);
        video.setAddedDate(addedDate);
        return video;
    }
}

File: killrvideo-service-suggestedvideo/src/test/kotlin/com/killrvideo/service/suggestedvideo/grpc/SuggestedVideosServiceGrpcMapperTest.kt
package com.killrvideo.service.suggestedvideo.grpc

import com.killrvideo.dse.dto.ResultListPage
import com.killrvideo.service.suggestedvideo.dto.Video
import com.killrvideo.service.suggestedvideo.grpc.SuggestedVideosServiceGrpcMapper.GetRelatedVideosRequestExtensions.parse
import com.killrvideo.utils.GrpcMappingUtils
import com.killrvideo.utils.GrpcMappingUtils.instantToTimeStamp
import com.killrvideo.utils.GrpcMappingUtils.uuidToUuid
import killrvideo.suggested_videos.getRelatedVideosRequest
import killrvideo.video_catalog.events.youTubeVideoAdded
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Test
import java.time.Instant
import java.util.*

internal class SuggestedVideosServiceGrpcMapperTest {
    private val mapper = SuggestedVideosServiceGrpcMapper()
    @Test
    fun testMapVideoAddedtoVideoDTO() {
        val v = video()
        val videoAdded = youTubeVideoAdded {
            videoId = uuidToUuid(v.videoid!!)
            addedDate = instantToTimeStamp(v.addedDate!!)
            userId = uuidToUuid(v.userid!!)
            name = v.name!!
            tags.addAll(v.tags!!)
            previewImageLocation = v.previewImageLocation!!
            location = v.location!!
        }
        val video = mapper.mapVideoAddedtoVideoDTO(videoAdded)
        assertEquals(v.videoid, video.videoid)
        assertEquals(v.userid, video.userid)
        assertEquals(v.addedDate, video.addedDate)
        assertEquals(v.name, video.name)
        assertEquals(v.tags!!.size, video.tags!!.size)
        assertEquals(v.previewImageLocation, video.previewImageLocation)
        assertEquals(v.location, video.location)
    }

    @Test
    fun testMapVideotoSuggestedVideoPreview() {
        val v = video()
        val preview = mapper.mapVideotoSuggestedVideoPreview(v)
        assertEquals(v.videoid.toString(), preview.videoId.value)
        assertEquals(v.userid.toString(), preview.userId.value)
        assertEquals(v.addedDate, GrpcMappingUtils.timestampToInstant(preview.addedDate))
        assertEquals(v.name, preview.name)
        assertEquals(v.previewImageLocation, preview.previewImageLocation)
    }

    @Test
    fun testParseGetRelatedVideosRequestData() {
        val videoid = UUID.randomUUID()
        val request = getRelatedVideosRequest {
            videoId = uuidToUuid(videoid)
            pageSize = 2
            pagingState = "Paging state"
        }
        val pojo = request.parse()
        assertEquals(videoid, pojo.videoid)
        assertEquals(2, pojo.pageSize)
        assertNotNull(pojo.pagingState)
        assertEquals("Paging state", pojo.pagingState)
    }

    @Test
    fun testMapToGetRelatedVideosResponse() {
        val v = video()
        val videoid = UUID.randomUUID()
        val resultPage = ResultListPage(listOf(v), "next page")
        val response = mapper.mapToGetRelatedVideosResponse(resultPage, videoid)
        assertEquals(videoid.toString(), response.videoId.value)
        assertEquals("next page", response.pagingState)
        assertEquals(1, response.videosCount)
    }

    private fun video(): Video {
        val videoid = UUID.randomUUID()
        val userid = UUID.randomUUID()
        val addedDate = Instant.now()
        val name = "Game"
        val tagList = Arrays.asList("tag1", "tag2")
        val previewLocation = "url"
        val location = "locationUrl"
        return Video(
            videoid = videoid,
            userid = userid,
            name = name,
            tags = HashSet(tagList),
            previewImageLocation = previewLocation,
            location=location,
            addedDate = addedDate
        )
    }
}
